source/trigger/iterator.go

```diff
@@ -0,0 +1,477 @@
+func (iter *Iterator) Next(ctx context.Context) (sdk.Record, error) {
+	switch {
+	case iter.snapshot != nil:
+		hasNext, err := iter.snapshot.HasNext(ctx)
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("snapshot has next: %w", err)
+		}
+
+		if hasNext {
+			return iter.snapshot.Next(ctx)
+		}
+
+		if err = iter.switchToCDCIterator(ctx); err != nil {
+			return sdk.Record{}, fmt.Errorf("switch to cdc iterator: %w", err)
+		}
+
+		fallthrough
+
+	case iter.cdc != nil:
+		hasNext, err := iter.cdc.HasNext(ctx)
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("cdc has next: %w", err)
+		}
+
+		if hasNext {
+			return iter.cdc.Next(ctx)
+		}
+
+		return sdk.Record{}, sdk.ErrBackoffRetry
+
+	default:
+		return sdk.Record{}, errors.New("not initialized iterator")
+	}
+}
+
+// Ack appends the last processed value to the slice to clear the tracking table in the future.
+func (iter *Iterator) Ack(_ context.Context, position sdk.Position) error {
+	pos, err := ParseSDKPosition(position)
+	if err != nil {
+		return fmt.Errorf("parse position: %w", err)
+	}
+
+	if pos.Mode == ModeCDC {
+		return iter.cdc.pushValueToDelete(pos.LastProcessedVal)
+	}
+
+	return nil
+}
+
+func (iter *Iterator) Teardown(ctx context.Context) error {
+	var err error
+
+	if iter.snapshot != nil {
+		err = iter.snapshot.Close()
+	}
+
+	if iter.cdc != nil {
+		err = multierr.Append(err, iter.cdc.Close())
+	}
+
+	// if the LastProcessedVal is nil - remove CDC tracking logic from DB
+	if iter.position.LastProcessedVal == nil {
+		err = multierr.Append(err, iter.removeCDC(ctx))
+	}
+
+	if err != nil {
+		return fmt.Errorf("teardown iterator: %w", err)
+	}
+
+	return nil
+}
+
+// populateKey queries the DB for the name of the primary key column for a table
+// and, if it exists, populates a key field in the iterator or returns an error.
+func (iter *Iterator) populateKey(ctx context.Context) error {
+	if iter.key != "" {
+		return nil
+	}
+
+	err := iter.conn.QueryRow(ctx, queryPrimaryKeySelect, iter.table).Scan(&iter.key)
+	if err != nil {
+		return fmt.Errorf("scan primary key: %w", err)
+	}
+
+	return nil
+}
+
+// populateColumns populates the iterator column with names of the table columns.
+func (iter *Iterator) populateColumns(ctx context.Context) error {
+	if len(iter.columns) > 0 {
+		return nil
+	}
+
+	rows, err := iter.conn.Query(ctx, queryColumnNamesSelect, iter.table)
+	if err != nil {
+		return fmt.Errorf("select names of the table columns: %w", err)
+	}
+	defer rows.Close()
+
+	colName := ""
+	for rows.Next() {
+		err = rows.Scan(&colName)
+		if err != nil {
+			return fmt.Errorf("scan column types rows: %w", err)
+		}
+
+		iter.columns = append(iter.columns, colName)
+	}
+
+	return nil
+}
+
+// switchToCDCIterator stops Snapshot and initializes CDC iterator.
+func (iter *Iterator) switchToCDCIterator(ctx context.Context) error {
+	err := iter.snapshot.Close()
+	if err != nil {
+		return fmt.Errorf("stop snaphot iterator: %w", err)
+	}
+
+	iter.snapshot = nil
```

Comment:
Wouldn't it be better to close and set the `snapshot` to `nil` _after_ the CDC is successfully initialized?

Reply:
Yeah, I think, in this case, the order doesn't actually matter. It would just look a bit safer to me. Nevermind,and good job! :)